File: targetFunctions/common/matlabEngineCaller_customLayers/tmwExample/example_single/codegen/lib/matlabEngineCaller_customLayers/matlabEngineCaller_customLayers.c
/*
 * Non-Degree Granting Education License -- for use at non-degree
 * granting, nonprofit, educational organizations only. Not for
 * government, commercial, or other organizational use.
 *
 * matlabEngineCaller_customLayers.c
 *
 * Code generation for function 'matlabEngineCaller_customLayers'
 *
 */

/* Include files */
#include "rt_nonfinite.h"
#include "matlabEngineCaller_customLayers.h"
#include "matlabEngineCaller_customLayers_emxutil.h"
#include "matlabCallFun.h"

/* Function Definitions */
void matlabEngineCaller_customLayers(const double params[8], double contrast,
  const emxArray_char_T *funcName, const emxArray_char_T *funcPath, double
  bulkIn, double bulkOut, emxArray_real_T *output, double sRough_data[], int
  sRough_size[2])
{
  emxArray_char_T *funName;
  int i;
  int loop_ub;
  static const char cv0[17] = { '[', 't', 'o', 't', 'a', 'l', ',', 'l', 'a', 'y',
    'e', 'r', 's', ']', ' ', '=', ' ' };

  emxArray_char_T *pathCall;
  static const char cv1[34] = { '(', 'p', 'a', 'r', 'a', 'm', 's', ',', 'b', 'u',
    'l', 'k', '_', 'i', 'n', ',', 'b', 'u', 'l', 'k', '_', 'o', 'u', 't', ',',
    'c', 'o', 'n', 't', 'r', 'a', 's', 't', ')' };

  static const char cv2[4] = { 'c', 'd', ' ', '\'' };

  double outp[300];
  static const char cv3[2] = { '\'', ';' };

  double nLayers;
  double b_params[8];
  unsigned int rowCount;
  int colCount;
  double d0;
  emxInit_char_T(&funName, 2);
  sRough_size[0] = 1;
  sRough_size[1] = 1;
  sRough_data[0] = params[0];

  /* Need to reserve some meory for the referencenced variables */
  i = funName->size[0] * funName->size[1];
  funName->size[0] = 1;
  funName->size[1] = 51 + funcName->size[1];
  emxEnsureCapacity((emxArray__common *)funName, i, (int)sizeof(char));
  for (i = 0; i < 17; i++) {
    funName->data[funName->size[0] * i] = cv0[i];
  }

  loop_ub = funcName->size[1];
  for (i = 0; i < loop_ub; i++) {
    funName->data[funName->size[0] * (i + 17)] = funcName->data[funcName->size[0]
      * i];
  }

  for (i = 0; i < 34; i++) {
    funName->data[funName->size[0] * ((i + funcName->size[1]) + 17)] = cv1[i];
  }

  emxInit_char_T(&pathCall, 2);
  i = pathCall->size[0] * pathCall->size[1];
  pathCall->size[0] = 1;
  pathCall->size[1] = 6 + funcPath->size[1];
  emxEnsureCapacity((emxArray__common *)pathCall, i, (int)sizeof(char));
  for (i = 0; i < 4; i++) {
    pathCall->data[pathCall->size[0] * i] = cv2[i];
  }

  loop_ub = funcPath->size[1];
  for (i = 0; i < loop_ub; i++) {
    pathCall->data[pathCall->size[0] * (i + 4)] = funcPath->data[funcPath->size
      [0] * i];
  }

  for (i = 0; i < 2; i++) {
    pathCall->data[pathCall->size[0] * ((i + funcPath->size[1]) + 4)] = cv3[i];
  }

  memcpy(&b_params[0], &params[0], sizeof(double) << 3);
  matlabCallFun(b_params, 8.0, &funName->data[0], &pathCall->data[0], bulkIn,
                bulkOut, contrast, outp, &nLayers);
  rowCount = 1U;
  colCount = 0;

  /* reshape the output to [layers * 3] array */
  emxFree_char_T(&pathCall);
  emxFree_char_T(&funName);
  if (nLayers >= 1.0) {
    i = output->size[0] * output->size[1];
    output->size[0] = (int)nLayers;
    output->size[1] = 3;
    emxEnsureCapacity((emxArray__common *)output, i, (int)sizeof(double));
    loop_ub = (int)nLayers * 3;
    for (i = 0; i < loop_ub; i++) {
      output->data[i] = 0.0;
    }

    d0 = nLayers * 3.0;
    for (i = 0; i < (int)d0; i++) {
      /* Make use of Matlab linear indexing. */
      output->data[((int)rowCount + output->size[0] * colCount) - 1] = outp[i];
      rowCount++;
      if (rowCount == nLayers + 1.0) {
        rowCount = 1U;
        colCount++;
      }
    }
  } else {
    i = output->size[0] * output->size[1];
    output->size[0] = 0;
    output->size[1] = 0;
    emxEnsureCapacity((emxArray__common *)output, i, (int)sizeof(double));
    sRough_size[0] = 0;
    sRough_size[1] = 0;
  }
}

/* End of code generation (matlabEngineCaller_customLayers.c) */

File: dev/matlabEngineCaller/matlabEngineCaller_customLayers.m
function [output,sRough] = matlabEngineCaller_customLayers(params,contrast,funcName,funcPath,bulkIn,bulkOut)

% Evaluate matlab function from compiled mex using an external Matlab
% engine. This version uses the Matlab Engine API for C 
% (https://uk.mathworks.com/help/matlab/calling-matlab-engine-from-c-programs-1.html)

% In reality, we want to be using a Class via the Matleb engine API for C++
% (https://uk.mathworks.com/help/matlab/matlab_external/matlab-engine-api-for-c.html)


% Irrespective of we use Matlab or Mex, we need the same calling string.
% We put together the Matlab command string that that we will excecute
paramsLen = int16(length(params));
funName = ['[total,layers] = ' funcName '(params,bulk_in,bulk_out,contrast)'];
sRough = params(1);

% Also make a commans string to change the Matlab Path to the location of
% the Matlab function
pathCall= ['cd ''' funcPath ''';'];

debug = 0;

% If not compiled, excecute our custom function from Matlab
if coder.target('MATLAB')
    if debug == 1
        % Show our command strings
        disp('Debug');
        disp(funName);
        disp(pathCall);
    end
    
    % If the file extensin is present, strip it off
    dotM = strfind(funcName,'.m');
    if ~isempty(dotM)
        funcName = funcName(1:dotM-1);
    end
    
    % Call the function using a handle
    fileHandle = str2func(funcName);
    [output,layers] = fileHandle(params,bulkIn,bulkOut,contrast);
else
    
    % Matlab coder will compile from here
   
    
    % Path to the engine 'h' files. This is always relative to the base
    % Matlab path on any system
    incPath1 = '/Applications/MATLAB_R2021a.app/extern/include';
    
    % Link path to the Matlab shared libraries. This will vary depending on
    % the OS
    linkPath1 = '/Applications/MATLAB_R2021a.app/bin/maci64';
    
    % The libraries we need
    linkFile1 = 'libeng.dylib';
    linkFile2 = 'libmx.dylib';
    
    % The source file including Path
    thisPath = '/Users/arwel/Documents/coding/RAT/dev/matlabEngineCaller';
    source1 = 'matlabCallFun.c';
    
    %source2 = '/home/arwel/Documents/RascalDev/RAT/targetFunctions/common/matlabEngineCaller_customLayers/matlabCallFun.h';
    
    %source1 = 'matlabCallFun.c';
    %source2 = 'matlabCallFun.h';
    
    libPriority = '';
    libPreCompiled = true;
    libLinkOnly = true;

    %coder.cinclude(source2);
    coder.updateBuildInfo('addSourceFiles',source1);
    %coder.updateBuildInfo('addSourceFiles',source2);
    coder.updateBuildInfo('addSourcePaths',thisPath);
    coder.updateBuildInfo('addIncludePaths',incPath1);
    %coder.updateBuildInfo('addIncludePaths',incPath2);
    coder.updateBuildInfo('addLinkObjects',linkFile1,linkPath1,libPriority,libPreCompiled,libLinkOnly);
    coder.updateBuildInfo('addLinkObjects',linkFile2,linkPath1,libPriority,libPreCompiled,libLinkOnly);
    
    %Need to reserve some meory for the referencenced variables
    outp = zeros(100,3);
    nLayers = 0;
    
    coder.ceval('matlabCallFun', params, paramsLen, funName, pathCall, bulkIn, bulkOut, contrast, coder.wref(outp), coder.wref(nLayers));
    
    rowCount = 1;
    colCount = 1;
    
    m = size(outp,1);
    n = size(outp,2);
    
    %reshape the output to [layers * 3] array
    if (nLayers >= 1)
        output = zeros(nLayers,3);
        for i = 1:(nLayers*3)
            thisVal = outp(i);  %Make use of Matlab linear indexing.
            output(rowCount,colCount) = thisVal;
            rowCount = rowCount + 1;
            if rowCount == (nLayers+1)
                rowCount = 1;
                colCount = colCount + 1;
            end
        end
    else
        output = [1 1 1];
        sRough = 1;
    end

end
